File: api-impl-py/tests/test_citation_conformance.py
"""Golden conformance: the Python apply_citations must produce IDENTICAL output
to the JS reference for the shared, language-neutral fixtures.

These are the same fixtures the JS impl runs (conformance/fixtures/citation-apply/).
This file is the proof that the port is behaviorally equivalent.
"""

import json
import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from irg.citations.apply import apply_citations  # noqa: E402

FIXTURE_DIR = os.path.normpath(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "conformance", "fixtures", "citation-apply")
)


class TestCitationConformance(unittest.TestCase):
    def test_fixtures_exist(self):
        files = [f for f in os.listdir(FIXTURE_DIR) if f.endswith(".json")]
        self.assertTrue(len(files) > 0, f"no fixtures in {FIXTURE_DIR}")

    def test_each_fixture_matches(self):
        files = sorted(f for f in os.listdir(FIXTURE_DIR) if f.endswith(".json"))
        for fname in files:
            with self.subTest(fixture=fname):
                with open(os.path.join(FIXTURE_DIR, fname), encoding="utf-8") as fh:
                    fixture = json.load(fh)
                result = apply_citations(fixture["draft_with_tags"], fixture["citable_set"])
                expected = fixture["expected"]

                self.assertEqual(result["prose"], expected["validated_prose"],
                                 f"{fname}: validated_prose mismatch")
                self.assertEqual(result["references"], expected["references"],
                                 f"{fname}: references mismatch")
                if "stats" in expected:
                    self.assertEqual(result["stats"], expected["stats"],
                                     f"{fname}: stats mismatch")
                # Stronger byte-comparability check: canonical JSON of references.
                self.assertEqual(
                    json.dumps(result["references"], sort_keys=True),
                    json.dumps(expected["references"], sort_keys=True),
                    f"{fname}: references not byte-comparable",
                )


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: api-impl-js/test/test-citation-conformance.js
/**
 * Golden conformance fixtures for citationApply (Citation_Application.md §10).
 *
 * Each fixture is a pure, language-neutral case:
 *   { draft_with_tags, citable_set } → { validated_prose, references[], stats }
 *
 * citationApply is deterministic, so these fixtures must diff exactly across
 * every polyglot port (JS / Python / Rust / LangChain). The citable_set
 * carries explicit uuids so this corpus tests the validation contract
 * independently of the uuid-derivation step (which has its own tests).
 */

'use strict';

const fs = require('fs');
const path = require('path');
const { applyCitations } = require('../core/citations/apply');

// Shared, language-neutral conformance corpus (consumed by every polyglot port).
const FIXTURE_DIR = path.join(__dirname, '..', '..', 'conformance', 'fixtures', 'citation-apply');

let passed = 0;
let failed = 0;
function ok(label, cond, extra) {
  if (cond) { passed++; console.log(`  ✓ ${label}`); }
  else      { failed++; console.log(`  ✗ ${label}`); if (extra !== undefined) console.log('    →', extra); }
}

function deepEqual(a, b) {
  return JSON.stringify(a) === JSON.stringify(b);
}

function main() {
  console.log('========================================');
  console.log('citationApply golden conformance fixtures');
  console.log('========================================');

  const files = fs.readdirSync(FIXTURE_DIR).filter((f) => f.endsWith('.json')).sort();
  ok('found fixtures', files.length > 0, files);

  for (const file of files) {
    const fixture = JSON.parse(fs.readFileSync(path.join(FIXTURE_DIR, file), 'utf8'));
    console.log(`\n[${file}] ${fixture.name}`);
    const result = applyCitations(fixture.draft_with_tags, fixture.citable_set);

    ok('validated_prose matches', result.prose === fixture.expected.validated_prose,
       `\n      expected: ${JSON.stringify(fixture.expected.validated_prose)}\n      actual:   ${JSON.stringify(result.prose)}`);
    ok('references match', deepEqual(result.references, fixture.expected.references),
       `\n      expected: ${JSON.stringify(fixture.expected.references)}\n      actual:   ${JSON.stringify(result.references)}`);
    if (fixture.expected.stats) {
      ok('stats match', deepEqual(result.stats, fixture.expected.stats),
         `\n      expected: ${JSON.stringify(fixture.expected.stats)}\n      actual:   ${JSON.stringify(result.stats)}`);
    }
  }

  console.log('\n========================================');
  console.log(`Results: ${passed} passed, ${failed} failed`);
  console.log('========================================');
  process.exit(failed === 0 ? 0 : 1);
}

main();
